core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java

```diff
@@ -202,6 +219,14 @@ else if (!splitWaitingForAnyNode) {
                     blockedExactNodes.addAll(candidateNodes);
                 }
             }
+
+            if (hasExhaustedNode) {
```

Comment:
If we have more splits to assign than possible queue size for all nodes, we want to quickly exit the assignment loop.

Reply:
One node is exhausted, but other might not be.

Reply:
If any node was exhausted, `randomCandidates` is being recreated without it if there are any nodes available.

Reply:
it was already a case for `        for (Split split : remainingSplits) {` loop that node could be exhausted.

Reply:
There are two situations I want to solve here:
1) when we choose random nodes from all nodes, we could take nodes that already are full and cannot take new split
2) if all nodes are full, we should end this loop, as it will not create a new assignment. 

That's why nodes are removed from `schedulableNodes` most of this situation could occur if number of nodes is smaller than `minCandidates` or number of splits exceed possible capacity of all nodes. In this case, we return `blockedFuture` and await it completion, which means that nodes have capacity for new assignments.